spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/MaterializingIterator.scala

```diff
@@ -0,0 +1,37 @@
+package com.datastax.spark.connector.util
+
+import java.util.NoSuchElementException
+
+import scala.collection.mutable
+
+/** Materializes/Fetches up to `materializationLimit` from underlying `iterator`.
+  * Materialization happens on `next()` invocation.
+  *
+  * In following example `it.next()` is invoked three times, results are caches in internal queue.
+  * {{{
+  * new MaterializingIterator(it, 3).next()
+  * }}}
+  *
+  */
+class MaterializingIterator[T](iterator: Iterator[T], materializationLimit: Int) extends Iterator[T] {
+
+  assert(materializationLimit > 0, "Materialization limit must be greater than 0")
+
+  private val materialized: mutable.Queue[T] = new mutable.Queue[T]()
+
+  private def materialize(): Unit = {
+    while (iterator.hasNext && materialized.size < materializationLimit) {
+      materialized.enqueue(iterator.next)
+    }
+  }
+
+  override def hasNext = !materialized.isEmpty || iterator.hasNext
+
+  override def next() = {
+    if(!hasNext) {
+      throw new NoSuchElementException("No more elements")
+    }
+    materialize()
```

Comment:
Not sure I understand the advantage here. 
This implementation will pay upfront the price of an initial n fetches but after that will behave like the underlying iterator until there're no more element in the backing iterator, when this wrapper will be faster b/c of the cached elements.

Is that what is intended? I guess that if we really want to amortize the underlying fetch for more than the initial `n` elements and to spread that profit across the iterating process, we will need some low watermark to trigger a materialization stage. 

WDYT?